internal/pkg/manifest/env.go

```diff
@@ -102,9 +109,107 @@ type environmentNetworkConfig struct {
-	ID      *string              `yaml:"id,omitempty"`
-	CIDR    *IPNet               `yaml:"cidr,omitempty"`
-	Subnets subnetsConfiguration `yaml:"subnets,omitempty"`
+	ID                  *string              `yaml:"id,omitempty"`
+	CIDR                *IPNet               `yaml:"cidr,omitempty"`
+	Subnets             subnetsConfiguration `yaml:"subnets,omitempty"`
+	SecurityGroupConfig securityGroupConfig  `yaml:"security_group,omitempty"`
+}
+
+type securityGroupConfig struct {
+	Ingress []SecurityGroupRule `yaml:"ingress,omitempty"`
+	Egress  []SecurityGroupRule `yaml:"egress,omitempty"`
+}
+
+func (cfg securityGroupConfig) isEmpty() bool {
+	return len(cfg.Ingress) == 0 && len(cfg.Egress) == 0
+}
+
+// SecurityGroupRule holds the security group ingress and egress configs.
+type SecurityGroupRule struct {
+	CidrIP      string      `yaml:"cidr"`
+	PortsConfig PortsConfig `yaml:"ports"`
+	IpProtocol  string      `yaml:"ip_protocol"`
+}
+
+// PortsConfig represents a range of ports [from:to] inclusive.
+// The simple form allow represent from and to ports as a single value, whereas the advanced form is for different values.
+type PortsConfig struct {
+	Port  *int            // 0 is a valid value, so we want the default value to be nil.
+	Ports *PortsRangeBand // Mutually exclusive with port.
+}
+
+// PortsRangeBand represents a range of from and to port values.
+// For example, "0-65536".
+type PortsRangeBand string
+
+// parse parses Ports string and returns the from and to port values.
+// For example: 1-100 returns	1 and 100.
+func (str PortsRangeBand) parse() (fromPort int, toPort int, err error) {
+	ports := strings.Split(string(str), "-")
+	fromPort, err = strconv.Atoi(ports[0])
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot convert from_port value %s to integer", ports[0])
+	}
+	toPort, err = strconv.Atoi(ports[1])
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot convert to_port value %s to integer", ports[1])
+	}
+	return fromPort, toPort, nil
+}
+
+// IsEmpty returns whether PortsRangeBand is empty.
+func (cfg *PortsRangeBand) IsEmpty() bool {
+	return cfg == nil || *cfg == ""
+}
+
+// IsEmpty returns whether PortsConfig is empty.
+func (cfg *PortsConfig) IsEmpty() bool {
+	return cfg.Port == nil && cfg.Ports.IsEmpty()
+}
+
+// Ports returns the from and to ports of a security group rule.
+func (r SecurityGroupRule) Ports() (from, to int, err error) {
+	if r.PortsConfig.Ports.IsEmpty() {
+		return *r.PortsConfig.Port, *r.PortsConfig.Port, nil // a single value is provided for ports.
+	}
+	return r.PortsConfig.Ports.parse()
+}
+
+// UnmarshalYAML overrides the default YAML unmarshaling logic for the Ports
+// struct, allowing it to perform more complex unmarshaling behavior.
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (cfg *PortsConfig) UnmarshalYAML(value *yaml.Node) error {
+	_, err := strconv.Atoi(value.Value) // detmines if input is integer? if err then we decode range values
```

Comment:
qq: is this check necessary? If the input isn't an integer, `value.Decode(&cfg.Port)` would have returned an error anyway, right?

Reply:
if customer somehow defines ports: "10" then it fails. Also when I write an integration test, `ports` are considered as string and then it unmarshals `ports: 10` into `PortsRangeBand` and the integ test fails. Hence converting

Reply:
What about swap the order, trying to unmarshal into int first then string?

Reply:
I tried doing that as well. But the problem is that if I unmarshal into int first then it sets the value of the for example `0-65335` as 0 in the int variable and then int variable becomes non empty and wrong unmarshaling happens. So that is why I am trying to convert input into int 1st and if it is not convertible then we know for sure that it's a range.